Return UNAVAILABLE on publish message and create workflow instance commands if no partitions are known

npepinpe opened the pull request:
## Description

This PR ensures we return UNAVAILABLE to clients on publish message and workflow instance creation commands if the gateway does not know any partitions. It also introduces a new system test for the gateway to check these cases, which can be extended to add more cases. 

## Related issues

closes #5783 
closes #5788 
closes #5789 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

Zelldon commented on gateway/src/main/java/io/zeebe/gateway/grpc/GrpcErrorMapper.java:
```diff
@@ -86,6 +87,10 @@ private Status mapErrorToStatus(final Throwable error, final Logger logger) {
+    } else if (error instanceof NoTopologyAvailableException) {
+      builder.setCode(Code.UNAVAILABLE_VALUE).setMessage(error.getMessage());
+      logger.trace(
+          "Expected to handle gRPC request, but the gateway does not know any partitions yet");
```
```suggestion
          "Expected to handle gRPC request, but the gateway does not know any partitions yet", error);
```

Zelldon commented on gateway/src/test/java/io/zeebe/gateway/api/UnavailableBrokersTest.java:
```diff
@@ -0,0 +1,95 @@
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+@Execution(ExecutionMode.CONCURRENT)
```
What does this?

npepinpe replied:
Tests are executed concurrently instead of one after the other. Should be used only when we are sure that they don't interfere between each other :slightly_smiling_face:

Zelldon replied:
thanks :)

Zelldon reviewed (approved):
No weekend for you ? :sweat_smile: 
Thanks @npepinpe for fixing this so quickly! :+1: I think https://github.com/zeebe-io/zeebe/issues/5702 is then also fixed or?

npepinpe commented:
> No weekend for you ? :sweat_smile:
Actually I was showing Zeebe to a friend, trying to do a little demo app with him, and I found a bunch of small issues surrounding the gateway :see_no_evil: 

> Thanks @npepinpe for fixing this so quickly! +1 I think #5702 is then also fixed or?
I don't think I fixed that, really - right now if I add a test case for deployment, I get DEADLINE_EXCEEDED. You can see it in `BrokerRequestManager:221` that we explicitly set the deployment partition if there's no topology, so yeah... But it should be easier to fix now, and extend with this test.

npepinpe commented:
btw, I wasn't sure to back port this because we didn't back port the GrpcErrorMapper and all this stuff, so it would be a pain to do so; if we want to backport this fix we'd need to backport that PR as well. Which imo we should do, but I could use a second opinion :slightly_smiling_face:

npepinpe commented:
bors r+

Zelldon commented:
I vote for backporting both
